src/Ocelot/Multiplexer/MultiplexingMiddleware.cs

```diff
@@ -1,219 +1,265 @@
-                            .SetIInternalConfiguration(httpContext.Items.IInternalConfiguration());
-
-                        newHttpContext.Items
-                            .UpsertTemplatePlaceholderNameAndValues(templatePlaceholderNameAndValues);
-
-                        newHttpContext.Items
-                            .UpsertDownstreamRoute(downstreamRoute);
-
-                        tasks.Add(Fire(newHttpContext, _next));
-                    }
-                }
+                tasksList.AddRange(ProcessRouteWithAggregation(matchAdvancedAgg, jObject, context, downstreamRoute));
+                continue;
+            }
 
-                await Task.WhenAll(tasks);
+            tasksList.Add(ProcessRouteAsync(context, downstreamRoute, _next));
+        }
 
-                var contexts = new List<HttpContext> { mainResponse };
+        return await Task.WhenAll(tasksList);
+    }
 
-                foreach (var task in tasks)
-                {
-                    var finished = await task;
-                    contexts.Add(finished);
-                }
+    /// <summary>
+    /// Mapping responses.
+    /// </summary>
+    private Task MapResponses(HttpContext context, Route route, HttpContext mainResponse,
+        IEnumerable<HttpContext> additionalResponses)
+    {
+        var contexts = new List<HttpContext> { mainResponse };
+        contexts.AddRange(additionalResponses);
+        return Map(context, route, contexts);
+    }
 
-                await Map(httpContext, route, contexts);
-            }
+    /// <summary>
+    /// Processing a route with aggregation.
+    /// </summary>
+    private IEnumerable<Task<HttpContext>> ProcessRouteWithAggregation(AggregateRouteConfig matchAdvancedAgg,
+        JToken jObject, HttpContext httpContext, DownstreamRoute downstreamRoute)
+    {
+        var tasks = new List<Task<HttpContext>>();
+        var values = jObject.SelectTokens(matchAdvancedAgg.JsonPath).Select(s => s.ToString()).Distinct();
+        foreach (var value in values)
+        {
+            var tPnv = httpContext.Items.TemplatePlaceholderNameAndValues();
+            tPnv.Add(new PlaceholderNameAndValue('{' + matchAdvancedAgg.Parameter + '}', value));
+            tasks.Add(ProcessRouteAsync(httpContext, downstreamRoute, _next, tPnv));
         }
 
-        private static HttpContext Copy(HttpContext source)
-        {
-            var target = new DefaultHttpContext();
+        return tasks;
+    }
 
-            foreach (var header in source.Request.Headers)
-            {
-                target.Request.Headers.TryAdd(header.Key, header.Value);
-            }
+    /// <summary>
+    /// Process a downstream route asynchronously.
+    /// </summary>
+    /// <returns>The cloned Http context.</returns>
+    private static Task<HttpContext> ProcessRouteAsync(HttpContext sourceContext, DownstreamRoute route,
+        RequestDelegate next, List<PlaceholderNameAndValue> placeholders = null)
```

Comment:
Keep required args on the one line, but the optional ones on the 2nd line.

Reply:
ok